main.cc

```diff
@@ -1480,6 +1480,19 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
             supervisor::notify("loading tablet metadata");
             ss.local().load_tablet_metadata().get();
 
+            // We do not support tablet re-sharding yet, see https://github.com/scylladb/scylladb/issues/16739.
+            // To avoid undefined behaviour due to violated assumptions, that
+            // each tablet has a valid shard replica, we check this assumption here, and refuse startup if violated.
+            for (const auto& [table_id, tmap] : ss.local().get_token_metadata_ptr()->tablets().all_tables()) {
+                tmap.for_each_tablet([host_id] (locator::tablet_id tid, const tablet_info& tinfo) {
+                    for (const auto& replica : tinfo.replicas) {
+                        if (replica.host == host_id && replica.shard >= smp::count) {
+                            throw std::runtime_error("Detected a tablet with invalid replica shard, reducing shard count with tablet-enabled tables is not yet supported. Replace the node instead.");
+                        }
+                    }
+                }).get();
+            }
```

Comment:
I think the check is good, but it shouldn't be open-coded in main but rather we should call a function from tablets.cc

Reply:
Move to `locator/tablets.{hh,cc}` in v2.